Libplanet/Crypto/BlsSignature.cs

```diff
@@ -0,0 +1,229 @@
+    /// <para> The signature is used for verification whether an <see cref="BlsPrivateKey"/> has any
+    /// relation with a message, by signing.</para>
+    /// The class allows to aggregate signatures and verifies aggregated signature,
+    /// <see cref="BlsSignature.AggregateVerify"/>,
+    /// <see cref="BlsSignature.FastAggregateVerify"/> or <see cref="BlsSignature.MultiVerify"/>
+    /// for its use-case.
+    /// <para>The implementation uses the BLS12_381 curve type with minimal public key
+    /// size setting, which is the same configuration with Ethereum 2.0 Phase 0.</para>
+    /// </summary>
+    /// <remarks>
+    /// The <see cref="BlsSignature"/> can be shared publicly.
+    /// The <see cref="BlsSignature"/> can be a single signature, aggregated signature, or
+    /// <see cref="BlsPrivateKey.ProofOfPossession"/>. Signature aggregation can be done by
+    /// <see cref="BlsSignatureExtensions.AggregateAll"/>.
+    /// <para>The infinite signature should be carefully treated, due to aggregating signatures with
+    /// infinite results an aggregated signature without infinite signature.</para>
+    /// <para>Every <see cref="BlsSignature"/> object is immutable.</para>
+    /// </remarks>
+    /// <seealso cref="Libplanet.Crypto.BlsPrivateKey"/>
+    /// <seealso cref="Libplanet.Crypto.BlsPublicKey"/>
+    public sealed class BlsSignature : IEquatable<BlsSignature>
+    {
+        private const int KeyByteSize = BLS.SIGNATURE_SERIALIZE_SIZE;
+        private readonly IReadOnlyList<byte> _signature;
+
+        /// <summary>
+        /// Instantiates a new <see cref="BlsSignature"/> by given <see cref="byte"/>
+        /// representation.
+        /// </summary>
+        /// <param name="signature">A <see cref="byte"/> representation of signature.</param>
+        /// <exception cref="ArgumentNullException">Thrown if given value is
+        /// <see cref="ImmutableArray{T}"/> and empty.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if given signature length is not
+        /// <see cref="KeyByteSize"/>.
+        /// </exception>
+        /// <exception cref="CryptographicException">Thrown if library has failed to deserializing
+        /// the signature (e.g., Invalid).</exception>
+        public BlsSignature(IReadOnlyList<byte> signature)
+        {
+            if (signature is ImmutableArray<byte> i ? i.IsDefaultOrEmpty : !signature.Any())
+            {
+                throw new ArgumentNullException(
+                    nameof(signature), "Signature is empty.");
+            }
+
+            if (signature.Count != KeyByteSize)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(signature),
+                    $"The key must be {KeyByteSize} bytes."
+                );
+            }
+
+            _signature = signature;
+            _ = CryptoConfig.ConsensusCryptoBackend.ValidateGetNativeSignature(ToByteArray());
+        }
```

Comment:
There is some difference between `IReadOnly` and `ImmutableArray` / `ImmutableList`. Either this should accept immutable type or copy `signature` passed in as argument. Also I think `_signature` should also be an immutable type.

Reply:
I've changed the type of `_signature` and other `_privatekey` and `_publickey` to `ImmutableArray<byte>`, and also `IPublic/PrivateKey.KeyBytes`